pkg/scheduler/schedule_one.go

```diff
@@ -843,17 +841,22 @@ func truncateMessage(message string) string {
 	return message[:max-len(suffix)] + suffix
 }
 
-func updatePod(client clientset.Interface, pod *v1.Pod, condition *v1.PodCondition, nominatingInfo *framework.NominatingInfo) error {
+func updatePod(client clientset.Interface, pod *v1.Pod, condition *v1.PodCondition, nominatingInfo *framework.NominatingInfo) {
 	klog.V(3).InfoS("Updating pod condition", "pod", klog.KObj(pod), "conditionType", condition.Type, "conditionStatus", condition.Status, "conditionReason", condition.Reason)
 	podStatusCopy := pod.Status.DeepCopy()
 	// NominatedNodeName is updated only if we are trying to set it, and the value is
 	// different from the existing one.
 	nnnNeedsUpdate := nominatingInfo.Mode() == framework.ModeOverride && pod.Status.NominatedNodeName != nominatingInfo.NominatedNodeName
 	if !podutil.UpdatePodCondition(podStatusCopy, condition) && !nnnNeedsUpdate {
-		return nil
+		return
 	}
 	if nnnNeedsUpdate {
 		podStatusCopy.NominatedNodeName = nominatingInfo.NominatedNodeName
 	}
-	return util.PatchPodStatus(client, pod, podStatusCopy)
+	go func() {
```

Comment:
We are using go-client's `retry.OnError` which uses a `wait.ConditionFunc` underneath.

> or when it was successful and you should not retry?

see the update in `pkg/scheduler/utils`

> or if you leak goroutines and create a new one despite the previous didn't finish?

We only retry 6 times (but probably will change to 4), is it that a big concern?

Reply:
> Can you elaborate?

When it comes to goroutines, it increases uncertainty. For the changes introduced here, I don't see a significant benefit that can outweigh the uncertainty.

If we agree it's a transient error, retry in a blocking manner makes more sense to me. While if the error is consistent, retry in goroutines would increase the burden to APIServer.

Reply:
if the errors we expect here are mostly api-server connectivity ones, then a blocking call is likely sufficient; the scheduler will not be able to make too much progress with the api-server down anyways.

Reply:
Okay, understood. If a new request to update a Pod (like bind request or another patch request from the next scheduling cycle) is sent from the scheduler (and the Pod is updated successfully) at a stage where the past one has not been completed by retries, the new Pod state, which is updated by the new request, may be overwritten by the past request. That's a problem..

> If we agree it's a transient error, retry in a blocking manner makes more sense. While if the error is consistent, retry in goroutines would increase the burden to APIServer.

> the scheduler will not be able to make too much progress with the api-server not reachable anyways.

These make sense. When api-server is unreachable from a scheduler, other requests are likely to fail as well, making it difficult for the scheduler to continue scheduling successfully. And putting more load on api-server by many requests from these goroutines is no solution.

I'll try to update the PR to change here to retry with a blocking.

Reply:
Oh, I didn't know that...  
> client-go implements some retry logic for requests it considers retryable, generally:
> - no write requests (PUT,POST) , a GET is retryable
> - transient network errors: connection reset by peer or EOF
> - retryable http errors: 50x code with a Retry-After header

So, we need to retry for "connection refused" by ourselves, but client-go do retries for 10s for errors that contain `Retry-After` like [`ServerTimeout`](https://github.com/kubernetes/kubernetes/blob/fbdd0d7b4165bc5a677d45e4dc693e3260297bfa/staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go#L363) 
[The current implementation](https://github.com/kubernetes/kubernetes/pull/109832/files#diff-6984d9ecf223c8282a6c2cde85aaea2cc72f6c3365881edca0d9f928ea0d7997R151-R162) does retry in addition to client-go does. Can we remove the retry in this case? What do you all think?

Reply:
Do we have to do anything to benefit from the builtin retries?

Is it the case that "connection refused" is not included?

Reply:
> Do we have to do anything to benefit from the builtin retries?

No. (right? @aojea)

> Is it the case that "connection refused" is not included in the retriable errors?

I guess this is also No because "connection refused" is not returned from api-server, and it isn't expected to have `Retry-After` header.